rir/src/compiler/opt/hoist_instruction.cpp

```diff
@@ -0,0 +1,187 @@
+                        } else {
+                            success = false;
+                        }
+                    }
+
+                    if (target == bb || !dom.dominates(target, bb))
+                        success = false;
+                });
+                if (!success || !target)
+                    target = nullptr;
+                else
+                    // both branches dominate bb, then we should move target
+                    // forward until they join again
+                    while (target->next0 && target->next0 != bb &&
+                           dom.dominates(target->next0, bb) &&
+                           (!target->next1 || dom.dominates(target->next1, bb)))
+                        target = target->next0;
+            }
+
+            if (!target) {
+                ip = next;
+                continue;
+            }
+
+            auto doesNotReorderEffects = [&](BB* x) {
+                std::function<bool(BB*)> compute = [&](BB* x) {
+                    if (!x || x == target)
+                        return true;
+                    // done, check cur bb for effects
+                    if (x == bb) {
+                        for (auto& j : *bb) {
+                            if (i == j)
+                                return true;
+                            if (j->hasStrongEffects())
+                                return false;
+                        }
+                        assert(false);
+                    }
+                    if (!x->isEmpty()) {
+                        // We can only hoist effects over branches if both
+                        // branch targets will trigger the effect
+                        if (x->last()->branches()) {
+                            if (!dom.dominates(x->trueBranch(), bb) ||
+                                !dom.dominates(x->falseBranch(), bb))
+                                return false;
+                        }
+                    }
+
+                    for (auto& j : *bb)
+                        if (j->hasStrongEffects())
+                            return false;
+
+                    return compute(x->next0) && compute(x->next1);
+                };
+                return compute(x->next0) && compute(x->next1);
+            };
+
+            auto noUnneccessaryComputation = [&](BB* x, unsigned exceptions) {
+                std::function<bool(BB*)> compute = [&](BB* x) {
+                    if (!x || x == bb || x == target)
+                        return true;
+                    if (!x->isEmpty()) {
+                        // If we hoist over a branch, but one of the
+                        // branches does not need the value, then this will
+                        // waste computation
+                        if (x->last()->branches()) {
+                            if (!dom.dominates(x->trueBranch(), bb) ||
+                                !dom.dominates(x->falseBranch(), bb)) {
```

Comment:
Why there is one exception?

Reply:
supposedly a loop is something like:

```
a:
  branch b c
b:
    ....
    goto a
c:
```

so to hoist an instruction from `...` to `a:` we need to cross one branch instructions where not both branch targets dominate the loop body.

this is basically a very localized formulation of, only hoist out of one loop.

Reply:
as I said in the comment, the better solution would be, to compute loop headers. then explicitly hoist instructions there. this would also allow us to hoist instructions without arguments. here I am just using the positions of the arguments of an instruction as a cheap heuristic of where it could be profitable to place an instruction.

maybe this could be achieved by combining the changes here with #362